File: solution/kotlin/week02-exercise2/src/main/kotlin/Book.kt
class Book private constructor(
    val title: String,
    val author: String,
    private var copies: Int
) {
    companion object {
        fun tryCreateBook(title: String?, author: String?, copies: Int): Book? {
            return if (title != null && author != null && copies >= 0) {
                Book(title, author, copies)
            } else {
                null
            }
        }
    }

    fun addCopies(additionalCopies: Int) {
        if (additionalCopies > 0) {
            copies += additionalCopies
        }
    }

    fun removeCopies(soldCopies: Int) {
        if (soldCopies in 1..copies) {
            copies -= soldCopies
        }
    }

    fun copies(): Int = copies
    fun hasCopies(): Boolean = copies > 0
}

File: solution/c#/week02-object-calisthenics/BookStore/Book.cs
namespace BookStore
{
    public class Book(string title, string author, int copies)
    {
        public string Title => title;
        public string Author => author;
        public int Copies => copies;
        public bool HasCopies => Copies > 0;

        public static Book? TryCreateBook(string? title, string? author, int copies)
            => InformationInvalid(title, author, copies)
                ? null
                : new Book(title!, author!, copies);

        private static bool InformationInvalid(string? title, string? author, int copies)
            => title is null
               || author == null
               || copies < 0;

        public void AddCopies(int additionalCopies)
        {
            if (additionalCopies > 0)
            {
                copies += additionalCopies;
            }
        }

        public void RemoveCopies(int soldCopies)
        {
            if (soldCopies > 0 && copies >= soldCopies)
            {
                copies -= soldCopies;
            }
        }
    }
}